src/admin/mod.rs

```diff
@@ -44,33 +39,73 @@ use crate::{config, signal};
 pub struct Builder {
     addr: SocketAddr,
     workload_info: WorkloadInformation,
-    ready: Readiness,
+    ready: Ready,
 }
 
 pub struct Server {
     addr: SocketAddr,
-    ready: Readiness,
+    ready: Ready,
     server: hyper::server::Builder<hyper::server::conn::AddrIncoming>,
     workload_info: WorkloadInformation,
 }
 
-#[derive(Clone, Debug)]
-pub struct Readiness(Arc<AtomicBool>);
+/// Ready tracks whether the process is ready.
+#[derive(Clone, Debug, Default)]
+pub struct Ready(Arc<Mutex<HashSet<String>>>);
+
+impl Ready {
+    pub fn new() -> Ready {
+        Ready(Default::default())
+    }
+
+    /// register_task allows a caller to add a dependency to be marked "ready".
+    pub fn register_task(&self, name: &str) -> BlockReady {
+        self.0.lock().unwrap().insert(name.to_string());
+        BlockReady {
+            parent: self.clone(),
+            name: name.to_string(),
+        }
+    }
+
+    pub fn is_ready(&self) -> bool {
+        self.0.lock().unwrap().len() == 0
+    }
+}
+
+/// BlockReady blocks readiness until it is dropped.
+pub struct BlockReady {
+    parent: Ready,
+    name: String,
+}
+
+impl BlockReady {
+    pub fn subtask(&self, name: &str) -> BlockReady {
+        self.parent.register_task(name)
+    }
+}
+
+impl Drop for BlockReady {
```

Comment:
If pretty much every `drop` is explicit, why do we need this vs some explicit "done" method or wrapping it in something like

```
fn subtask(&self, name: &str, f: SomeFnType) {
  // add to set
  f()
  // remove
}

```

Reply:
The closure won't work great some other cases (in XDS its going to be deeply nested when we process the first request fully). We can remove the drop but then it makes it possible to forget about it and block readiness indefinitely